### PR TITLE
Pulp trainlib dev

### PR DESCRIPTION
**Pull request by Dequino**

### CODE REVIEW

File: lib/include/pulp_mhsa_fp32.h [hunk 1]
```diff
@@ -74,14 +74,14 @@ struct Mhsa_args {
...
  */
 void pulp_mhsa_fp32_fw_cl(void * Mhsa_args);
 
 
 /**
- * @brief Forward pass function, forked on PULP cluster, using partial softmax.
+ * @brief Forward pass function, forked on PULP cluster, using partial softmax. (allows parallelization, parallelize with pi_cl_team_fork(NUM_CORES, pulp_mhsa_fp32_fw_cl_2, &args)).
  * @param Mhsa_args structure configuring the MHSA layer.
  */
 void pulp_mhsa_fp32_fw_cl_2(void * Mhsa_args);
```
Comment (dnadalini) [inline lib/include/pulp_mhsa_fp32.h:87]: Perchè si chiama 2?

File: lib/sources/pulp_train_utils_fp16.c [hunk 2]
```diff
@@ -463,7 +463,7 @@ void pulp_scalar_mul_fp16_cl(void* void_args){
     struct scalar_mul_args_fp16* args = (struct scalar_mul_args_fp16 *) void_args;
 
     fp16* input = args->input;
-    fp16 scalar = args->scalar;
+    float scalar = args->scalar;
```
Comment (dnadalini) [inline lib/sources/pulp_train_utils_fp16.c:466]: Are you sure this is a float and not an fp16?

File: tests/test_conv_pw_dw_fp16/Makefile
Comment (dnadalini) [inline tests/test_conv_pw_dw_fp16/Makefile:1]: Remove from commit (all file)

File: tests/test_conv_pw_dw_fp16/net.h
Comment (dnadalini) [inline tests/test_conv_pw_dw_fp16/net.h:1]: Remove from commit (all file)

File: tests/test_conv_pw_dw_fp16/utils/GM.py
Comment (dnadalini) [inline tests/test_conv_pw_dw_fp16/utils/GM.py:1]: Remove from commit (all file)

File: tests/test_matmul/net.c [hunk 3]
```diff
@@ -145,6 +157,8 @@ static inline void multiply ()
     compare_tensors(result, C, IN_CH*OUT_CH);
     null_tensor(result, IN_CH*OUT_CH);
 
+    /*
+
     printf("\n-----> Profiling mm_unroll_8x1:\n");
```
Comment (dnadalini) [inline tests/test_matmul/net.c:162]: Is this under fix?

File: tests/test_matmul/net.c [hunk 4]
```diff
@@ -161,14 +175,6 @@ static inline void multiply ()
...
 
-    printf("\n-----> Profiling mm_unroll_2x4:\n");
-    START_STATS();
-    pi_cl_team_fork(NUM_CORES, mm_unroll_2x4, &mm_args);
-    STOP_STATS();
-    check_tensor(result, C, IN_CH*OUT_CH);
-    compare_tensors(result, C, IN_CH*OUT_CH);
-    null_tensor(result, IN_CH*OUT_CH);
-
     printf("\n-----> Profiling mm_unroll_4x2:\n");
```
Comment (dnadalini) [inline tests/test_matmul/net.c:178]: Is this under fix?

File: lib/include/pulp_mhsa_fp32.h (hunk 1 above)
Comment (Dequino) [reply lib/include/pulp_mhsa_fp32.h:87]: It's the legacy version of the mhsa kernel, I'll probably need it for comparison of my own work in a future journal paper (maybe)

File: lib/sources/pulp_train_utils_fp16.c (hunk 2 above)
Comment (Dequino) [reply lib/sources/pulp_train_utils_fp16.c:466]: yes

File: lib/include/pulp_linear_fp16.h [hunk 5]
```diff
@@ -94,4 +94,4 @@ void pulp_linear_fp16_bw_param_grads_cl( void * Linear_args_fp16 );
  * @param output  categorical output for the linear layer (from forward perspective)
  * @param opt_matmul_type_ig number of the optimizer matmul to be chosen by the mm_manager (see mm_manager_list.txt)
  */
-void pulp_linear_fp16_bw_input_grads_cl( void * Linear_args_fp16 );
```
[the remark this answers is not in the document]
Comment (Dequino) [reply lib/include/pulp_linear_fp16.h:97]: I can't.

File: tests/test_matmul/net.c (hunk 3 above)
Comment (Dequino) [reply tests/test_matmul/net.c:162]: yes, this has been fixed and needs to be like this

File: lib/include/pulp_mhsa_fp32.h (hunk 1 above)
Comment (dnadalini) [reply lib/include/pulp_mhsa_fp32.h:87]: Ok! Can you please change the name maybe in the next commit so that it is more clear what it is? Thnx!

**Review by dnadalini — APPROVED** [review]

**Pull request merged**